src/pow_fpga_accel.c

```diff
@@ -0,0 +1,156 @@
+/* 
+ * Copyright (C) 2018 dcurl Developers. 
+ * Copyright (c) 2018 Ievgen Korokyi. 
+ * Use of this source code is governed by MIT license that can be 
+ * found in the LICENSE file.
+ */
+
+#include "pow_fpga_accel.h"
+
+#define NONCE_LEN 81
+#define TRANSACTION_LEN 2673
+#define NONCE_OFFSET 2646
+#define HPS_TO_FPGA_BASE 0xC0000000
+#define HPS_TO_FPGA_SPAN 0x0020000
+#define HASH_CNT_REG_OFFSET 4
+#define TICK_CNT_LOW_REG_OFFSET 5
+#define TICK_CNT_HI_REG_OFFSET 6
+#define MWM_MASK_REG_OFFSET 3
+#define CPOW_BASE 0
+
+static FILE *ctrl_fd;
+static FILE *in_fd;
+static FILE *out_fd;
+static int devmem_fd;
+static void *fpga_regs_map;
+static uint32_t *cpow_map;
+
+int pow_fpga_accel_init()
+{
+    ctrl_fd = 0;
+    in_fd = 0;
+    out_fd = 0;
+    devmem_fd = 0;
+    fpga_regs_map = 0;
+    cpow_map = 0;
+
+    ctrl_fd = fopen("/dev/cpow-ctrl", "r+");
+
+    if (ctrl_fd == NULL) {
+        perror("cpow-ctrl open fail");
+        return 0;
+    }
+
+    in_fd = fopen("/dev/cpow-idata", "wb");
+
+    if (in_fd == NULL) {
+        perror("cpow-idata open fail");
+        fclose(ctrl_fd);
+        return 0;
+    }
+
+    out_fd = fopen("/dev/cpow-odata", "rb");
+
+    if (out_fd == NULL) {
+        perror("cpow-odata open fail");
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        return 0;
+    }
+
+    devmem_fd = open("/dev/mem", O_RDWR | O_SYNC);
+
+    if (devmem_fd < 0) {
+        perror("devmem open");
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        fclose(out_fd);
+        return 0;
+    }
+
+    fpga_regs_map =
+        (uint32_t *) mmap(NULL, HPS_TO_FPGA_SPAN, PROT_READ | PROT_WRITE,
+                          MAP_SHARED, devmem_fd, HPS_TO_FPGA_BASE);
+    cpow_map = (uint32_t *) (fpga_regs_map + CPOW_BASE);
+
+    if (fpga_regs_map == MAP_FAILED) {
+        perror("devmem mmap");
+        close(devmem_fd);
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        fclose(out_fd);
+        return 0;
+    }
+
+    return 1;
+}
+
+void pow_fpga_accel_destroy()
+{
+    int result;
+    
+    fclose(in_fd);
+    fclose(out_fd);
+    fclose(ctrl_fd);
+
+    result = munmap(fpga_regs_map, HPS_TO_FPGA_SPAN);
+
+    close(devmem_fd);
+
+    if (result < 0) {
+        perror("devmem munmap");
+        exit(EXIT_FAILURE);
```

Comment:
It seems that we need to change the return type so that the destroy function can return the status rather than exit(3).
But anyway, exit(3) should not be used because it may induce the IRI not correctly terminated. Or any reason of the use of exit(3) ?

Reply:
No, I only wanna fail to terminate the program.  I will drop it.